### ISSUE TITLE
Replace dance with Evaluator classes and raw calls to scipy with tensorflow.contrib.opt.ScipyOptimizerInterface

### ISSUE THREAD

**Issue by hnarayanan** [hnarayanan/Stylist#1]
We can perhaps get away without explicitly constructing gradients, and instead just tell the program: "Here is the loss, here are the variables you can change, here is the optimisation algorithm I'd like you to use," and have it automatically do the rest.

Helpful links:
- https://github.com/tensorflow/tensorflow/issues/446
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/external_optimizer.py#L246
